File: src/main/java/me/centralhardware/znatoki/telegram/statistic/eav/types/Integer.java
package me.centralhardware.znatoki.telegram.statistic.eav.types;

import io.vavr.control.Validation;
import org.apache.commons.lang3.StringUtils;
import org.telegram.telegrambots.meta.api.objects.Update;

public non-sealed class Integer implements Type {
    @Override
    public String format(String name, Boolean isOptional) {
        return STR."Введите \{name} (число). \{isOptional? optionalText: ""}";
    }

    @Override
    public Validation<String, Void> __validate(Update update, String...variants) {
        return update.hasMessage() && StringUtils.isNumeric(update.getMessage().getText()) ?
                Validation.valid(null) :
                Validation.invalid("Введите число");
    }

    @Override
    public String extract(Update update) {
        return update.getMessage().getText();
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/eav/types/Integer.kt
package me.centralhardware.znatoki.telegram.statistic.eav.types

import arrow.core.Either
import dev.inmo.tgbotapi.extensions.utils.extensions.raw.text
import dev.inmo.tgbotapi.types.message.abstracts.CommonMessage
import dev.inmo.tgbotapi.types.message.content.MessageContent
import me.centralhardware.znatoki.telegram.statistic.eav.types.Type.Companion.OPTIONAL_TEXT
import org.apache.commons.lang3.StringUtils

typealias NumberType = Integer

data object Integer : Type {

    override fun format(name: String, isOptional: Boolean): String {
        return "Введите $name (число). ${if (isOptional) OPTIONAL_TEXT else ""}"
    }

    override fun validate(
        message: CommonMessage<MessageContent>,
        variants: List<String>
    ): Either<String, Unit> {
        return if (StringUtils.isNumeric(message.text)) Either.Right(Unit)
        else Either.Left("Введите число")
    }
}
